### ISSUE TITLE
How does count() work?

### ISSUE THREAD

**Issue by andreas-zeller** [rindPHI/isla#39]
The current description of `count()` in ISLa lacks an example. How would I use it within a constraint?

Background: I am trying to have at least 3 (or exactly 3) nodes of a particular nonterminal in the generated output and wondered how I can achieve this.

**Comment by andreas-zeller** [comment]
Okay, I found an example under "numeric quantifiers" and now use

```
exists int seqs: (
    count(<start>, "<sequence>", seqs) and
    str.to.int(seqs) >= 3
    )
```

to ensure I have at least 3 sequences.

**Comment by rindPHI** [comment]
> The current description of `count()` in ISLa lacks an example. How would I use it within a constraint?
> 
> Background: I am trying to have at least 3 (or exactly 3) nodes of a particular nonterminal in the generated output and wondered how I can achieve this.

Where did you look for a description? The `count` predicate is described in the [ISLa Language Specification](https://rindphi.github.io/isla/islaspec/#semantic-predicates), which is also referred in the README. What are your suggestions as to where place a hint in addition?

**Comment by rindPHI** [comment]
> Okay, I found an example under "numeric quantifiers" and now use
> 
> ```
> exists int seqs: (
>     count(<start>, "<sequence>", seqs) and
>     str.to.int(seqs) >= 3
>     )
> ```
> 
> to ensure I have at least 3 sequences. But then, this produces a `StopIteration` exception, implying there's no solution,.

You probably hit a but related to the interplay of `count` with epsilon productions (can you confirm this by also posting your grammar?). I'm discussing the problem and a workaround in [this Jupyter notebook](https://github.com/rindPHI/isla/issues/39#issuecomment-1368008625). There's a workaround included.

**Comment by rindPHI** [comment]
Note: The lines

```python
if (
    not candidate_more_needles_possible
    and candidate_needle_occurrences == target_num_needle_occurrences
):
    return SemPredEvalResult({in_tree: candidate})
```

in `isla_predicate.py` should be updated such that we can replace nullable nonterminals by epsilons if the target number of "needles" has been reached.

**Comment by rindPHI** [comment]
Fixed in #42
